### PR TITLE
fix(grok-build): align LLM token usage & system_instructions array shape (F1/F3a)

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## 背景

承接 PR #159（F2/F4/F6/F8 semconv 对齐，OPEN），补齐 architect `1f3fb94f` APPROVED 的 P0 字段缺口：
- **F1**：LLM span `gen_ai.usage.input_tokens`/`output_tokens`/`cache_read` 真值对齐 `~/.grok/logs/unified.jsonl` `shell.turn.inference_done` 的 `prompt_tokens`/`completion_tokens`/`cached_prompt_tokens`（grok 0.2.101 transcript 的 `assistant.usage` 全 0，token=0 known gap）
- **F3a**：AGENT span `gen_ai.system_instructions` 按 semconv 数组形式 `[{type:'text', content}]` 发射（architect `1f3fb94f` APPROVED，`e9f78ef4` 修订 array 形状）

### F1 — `loadUsageBySession(sid)` (`assets/hooks/grok-build-hook-processor.mjs`)
- 读 `~/.grok/logs/unified.jsonl`，过滤 `msg=='shell.turn.inference_done' && sid==session_id && ctx.prompt_tokens!=null`（R1 retry null-token 跳过），按 `ts`+`loop_index` 排序
- `MAX_UNIFIED_BYTES=50MB` tail-read 防护（R2，324KB/21 sessions 膨胀，参照 `transcript-parser.mjs:121` `MAX_TRANSCRIPT_BYTES` 模式）
- `cmdStop` 调 `loadUsageBySession` → `exportSession` 用 `usageState` 游标（持久化 `state.usage_events_consumed`），按 session 顺序为每个 LLM call 弹一条 inference_done
- grok `prompt_tokens` 是 OpenAI 风格（total input，含 cached）→ `input_tokens=prompt_tokens`，`cache_read.input_tokens=cached_prompt_tokens`（不重复加，避免 double-count）
- 缺 `unified.jsonl` 时优雅 fallback 到 transcript `assistant.usage`（通常 0）

### F3a — `gen_ai.system_instructions` 数组
- `parseGrokTranscript` 返回增 `systemPrompt`（首个 `type:system` record content）(`assets/hooks/grok-build/transcript-parser.mjs`)
- AGENT span `baseFields` 首次 turn（`baseTurnCount+i===0`）注入 `gen_ai.system_instructions: [{type:'text', content: systemPrompt}]`，仅在 AGENT span 一次
- `systemPrompt` 持久化 `state.system_prompt`，增量 `cmdStop` 仍可取（增量解析时 system record 已在 offset 之后）

## 修改文件

| 文件 | 类型 | 改动 |
|------|------|------|
| `assets/hooks/grok-build/transcript-parser.mjs` | modified | +11/-8，F3a part 1：`parseGrokTranscript` 返回 `systemPrompt` |
| `assets/hooks/grok-build-hook-processor.mjs` | modified | +156/-8，F1 `loadUsageBySession` + F3a `gen_ai.system_instructions` 数组 + `MAX_UNIFIED_BYTES` 防护 |
| `tests/unit/hooks/grok-build/usage-injection.test.mjs` | new | 4 用例：3 LLM + 5 inference_done（3 非空 + 2 null-retry）按序注入 / retry null-token 跳过 / 缺 unified.jsonl fallback / session 游标跨 cmdStop 持久化 |
| `tests/unit/hooks/grok-build/system-instructions.test.mjs` | new | 4 用例：数组形状（Array.isArray/length===1/[0].type==='text'/content===fixture 全文）/ 多 turn 跨 cmdStop batch 首次 AGENT span 发射、turn 2+ 不发射 / 缺 system record 不崩溃 |
| `tests/unit/hooks/grok-build/fixtures/f3-system-prompt.txt` | new | 5077B fixture，来源 researcher `47a8af1f` 附件 `research-fixtures.tar.gz:f3-system-prompt.txt`（real grok 0.2.101 system_prompt.txt） |

## 验证结果

- **单测**：`npx vitest run tests/unit/hooks/grok-build/` → **37/37 PASS**（transcript-parser 12 + usage-injection 4 + system-instructions 4 + hook-processor 17）
- **E2E**：tester `e1cfa237` verdict **PASS** — 真实 grok 0.2.101 5 步 ReAct，6 STEP/6 LLM/5 TOOL，F1 token 与 unified.jsonl 完全对齐（loop 1-4 input/output/cache_read 全 ✅），F3a 数组形状全维度通过
- **validate-trace**：主场景 **0 ERROR**（仅 SHOULD 级 WARN，不在 F1/F3a 范围）
- **architect** `1f3fb94f` APPROVED

## 栈式 PR 说明

- **base**: `agent/pilot-reviewer-v2/grok-build-semconv-align`（PR #159 分支），非 `main`
- **diff vs PR #159**: +167/-16 modified + 677 new = 仅 F1/F3a delta，无 F2/F4/F6/F8 重复
- **PR #159 合并后 GitHub 自动 retarget base 到 main**（GitHub 默认行为：base 分支删除/合并后自动切换到默认分支）

## 上游引用

- developer `7b03c4e0`：F1/F3a 实现 + 37/37 单测
- tester `e1cfa237`：CP5 E2E PASS + validate-trace 0 ERROR
- architect `1f3fb94f`：APPROVED
- researcher `47a8af1f`：CP1 fixture 齐全
- coordinator `f1f9cd26`：裁决选 (A) 栈式 PR

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ralf0131** [comment]
**CLA Not Signed**

The Contributor License Agreement (CLA) check is currently pending on this PR (`license/cla`: Contributor License Agreement is not signed yet.). This PR **cannot be merged** until the CLA is signed.

@rangemer333-cell please sign the CLA via the CLA assistant badge in the comment above, or visit https://cla-assistant.io/alibaba/loongsuite-pilot. Once signed, the `license/cla` status will turn green.

---
*Automated check by github-manager-bot*

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Implements F1 (real LLM token usage injection from `~/.grok/logs/unified.jsonl`) and F3a (`gen_ai.system_instructions` spec-compliant array form) for the grok-build hook processor. Adds a session-level usage cursor (`state.usage_events_consumed`) persisted across cmdStop batches, 50MB tail-read protection on the unified log, and graceful fallback to transcript `assistant.usage` (typically 0 in grok 0.2.x). Both features include comprehensive test coverage (4 test cases each). Code quality is good — error handling, edge cases, and incremental parse support are well thought out.

**However, this PR cannot be approved/merged until the contributor signs the CLA** (current status: pending). See the CLA reminder above. Once CLA is signed, this looks ready to merge.

### Findings
- **[Warning]** `grok-build-hook-processor.mjs:470` — Cursor alignment edge case with `lastTurnOnly` (see inline comment)
- **[Info]** `grok-build-hook-processor.mjs:159` — `reasoning_tokens` captured but unused (see inline comment)

### Suggestions
- Consider advancing the usage cursor for all parsed turns (not just exported ones) to keep `usage_events_consumed` accurate across `lastTurnOnly` batches.
- Stacked on PR #159 — ensure merge order (#159 before #160) is documented for maintainers.

---
*Automated review by github-manager-bot*

File: assets/hooks/grok-build-hook-processor.mjs [hunk 1]
```diff
@@ -79,6 +79,87 @@ function defaultLogDir() {
...
+  out.sort((a, b) => {
+    if (a.ts && b.ts) {
+      const c = a.ts.localeCompare(b.ts);
+      if (c !== 0) return c;
+    }
+    if (a.loop_index != null && b.loop_index != null) {
+      return a.loop_index - b.loop_index;
+    }
+    return 0;
+  });
```
Comment (ralf0131) [inline assets/hooks/grok-build-hook-processor.mjs:159]: **[Info]** `reasoning_tokens` is captured in the `loadUsageBySession` output objects but never consumed in `buildTurnRecords`. If this field is reserved for a future semconv addition, a brief inline comment noting that would help future readers; otherwise, removing it keeps the data flow clean.

File: assets/hooks/grok-build-hook-processor.mjs [hunk 2]
```diff
@@ -358,12 +440,34 @@ async function exportSession(state, stopReason, fallbackTs) {
...
+  // call only if previous batches also advanced. To stay correct across batches,
+  // track a counter in state of how many inference_done events have been consumed.
+  const consumedSoFar = state.usage_events_consumed || 0;
+  const usageState = { events: usageEvents, idx: consumedSoFar };
+
   for (let i = 0; i < turnsToExport.length; i++) {
     const turn = turnsToExport[i];
     const isLast = i === turnsToExport.length - 1;
     const turnStopReason = isLast ? stopReason : 'end_turn';
+    const isAgentSpanTurn = (baseTurnCount + i) === 0;
```
Comment (ralf0131) [inline assets/hooks/grok-build-hook-processor.mjs:470]: **[Warning]** Cursor alignment edge case: when `lastTurnOnly` is true and multiple turns are parsed, only the last turn is exported via `buildTurnRecords`, so `usageState.idx` advances only for that turn's LLM calls. The skipped turns' inference_done events are not popped, leaving the persisted `usage_events_consumed` short. Since enrichment is explicitly best-effort (falls back to 0), this is non-blocking, but consider advancing the cursor for all parsed turns even when only the last is exported, or documenting this limitation in the comment.

**Comment by fangxiu-wf** [comment]
Closing: net delta of #160 has been squashed into #159 (single consolidated PR, commit 960403b, author fangxiu-wf). See https://github.com/alibaba/loongsuite-pilot/pull/159
